TelepresenceBot/mobile/src/main/java/com/novoda/tpbot/human/HumanSocketIOTpService.java

```diff
@@ -0,0 +1,117 @@
+class HumanSocketIOTpService implements HumanTpService {
+
+    private Socket socket;
+    private final Handler handler;
+
+    static HumanSocketIOTpService getInstance() {
+        return LazySingleton.INSTANCE;
+    }
+
+    private HumanSocketIOTpService() {
+        this.handler = new Handler(Looper.getMainLooper());
+    }
+
+    @Override
+    public Observable<Result> connectTo(String serverAddress) {
+        try {
+            URL url = new URL(serverAddress);
+            socket = IO.socket(url.toExternalForm());
+        } catch (MalformedURLException | URISyntaxException exception) {
+            String message = "Address should be in the format `http://[ip_address]:[port_number]`";
+            HumanSocketIOConnectionException exceptionWithUserFacingMessage = new HumanSocketIOConnectionException(message, exception);
+            return Observable.just(Result.from(exceptionWithUserFacingMessage));
+        }
+        return new SocketConnectionObservable();
+    }
+
+    private class SocketConnectionObservable extends Observable<Result> {
+
+        @Override
+        public Observable<Result> start() {
+            if (socket.connected()) {
+                notifyOf(Result.from("Already connected"));
+                return this;
+            }
+
+            socket.emit("join_as_human", "", joinSuccessAcknowledgementListener);
+            socket.on(EVENT_CONNECT_ERROR, connectionTimeoutListener);
+            socket.connect();
+            return this;
+        }
+
+        private final Ack joinSuccessAcknowledgementListener = new Ack() {
+            @Override
+            public void call(final Object... args) {
+                handler.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (args[0] != null) {
+                            Object object = args[0];
+                            notifyOf(Result.from(object.toString()));
+                        }
+                    }
+                });
+            }
+        };
+
+        private final Emitter.Listener connectionTimeoutListener = new Listener() {
+            @Override
+            public void call(Object... args) {
+                handler.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        notifyOf(Result.from(new SocketIOException("connection timeout")));
+                    }
+                });
+            }
+        };
+    }
+
+    @Override
+    public void moveIn(Direction direction) {
+        socket.emit("move_in", direction);
+    }
+
+    @Override
+    public void disconnect() {
+        if (socket != null) {
+            socket.disconnect();
+            socket.off();
+        }
+    }
+
+    private static class LazySingleton {
+        private static final HumanSocketIOTpService INSTANCE = new HumanSocketIOTpService();
+    }
+
+    public class HumanSocketIOConnectionException extends RuntimeException {
```

Comment:
does this have to be a runtime exception? also, given it's a subclass what about naming it `ConnectionException`?